File: app/src/main/java/de/synyx/android/meetingroom/domain/Agenda.java
package de.synyx.android.meetingroom.domain;

import android.support.annotation.NonNull;

import de.synyx.android.meetingroom.config.Registry;
import de.synyx.android.meetingroom.util.TimeProvider;

import io.reactivex.functions.BiFunction;

import org.joda.time.Duration;

import java.util.ArrayList;
import java.util.List;

import static de.synyx.android.meetingroom.domain.TimeSpan.zeroTimeSpanFrom;

import static io.reactivex.Observable.fromIterable;

import static java.util.Collections.sort;
import static java.util.Collections.unmodifiableList;


/**
 * @author  Max Dobler - dobler@synyx.de
 */
public class Agenda {

    private TimeProvider timeProvider = Registry.get(TimeProvider.class);

    private List<Reservation> reservations = new ArrayList<>();

    public void addReservation(Reservation reservation) {

        this.reservations.add(reservation);
        sort(reservations);
    }


    public Duration getTimeUntilNextMeeting() {

        return
            fromIterable(reservations) //
            .filter(Reservation::isUpcoming) //
            .firstElement() //
            .map(Reservation::getTimeUntilBegin) //
            .blockingGet();
    }


    public List<Reservation> getReservations() {

        return unmodifiableList(reservations);
    }


    public boolean hasCurrentReservation() {

        return fromIterable(reservations) //
            .any(Reservation::isCurrentlyActive) //
            .blockingGet();
    }


    public Duration getTimeUntilAvailable() {

        if (!hasCurrentReservation()) {
            return Duration.ZERO;
        }

        return
            fromIterable(reservations) //
            .filter(Reservation::isCurrentOrUpcoming) //
            .map(Reservation::toTimeSpan) //
            .reduce(zeroTimeSpanFrom(timeProvider.now()), sumTimeOfOngoingReservations()) //
            .map(TimeSpan::toDuration) //
            .blockingGet();
    }


    @NonNull
    private BiFunction<TimeSpan, TimeSpan, TimeSpan> sumTimeOfOngoingReservations() {

        return
            (available, nextTimeSpan) -> {
            if (nextTimeSpan.begin.isAfter(available.end)) {
                return available;
            }

            return new TimeSpan(available.begin, nextTimeSpan.end);
        };
    }


    public boolean isNextReservationInNearFuture() {

        if (hasCurrentReservation()) {
            return false;
        }

        Duration timeUntilNextMeeting = getTimeUntilNextMeeting();

        if (timeUntilNextMeeting == null) {
            return false;
        }

        return timeUntilNextMeeting.getStandardMinutes() <= 15L;
    }


    public Reservation getCurrentMeeting() {

        return fromIterable(reservations) //
            .filter(Reservation::isCurrentlyActive) //
            .firstElement() //
            .blockingGet();
    }


    public Reservation getUpcomingReservation() {

        return fromIterable(reservations) //
            .filter(Reservation::isUpcoming) //
            .firstElement() //
            .blockingGet();
    }
}

File: app/src/main/java/de/synyx/android/meeroo/domain/Agenda.kt
package de.synyx.android.meeroo.domain

import de.synyx.android.meeroo.config.Registry
import de.synyx.android.meeroo.domain.TimeSpan.zeroTimeSpanFrom
import de.synyx.android.meeroo.util.TimeProvider
import org.joda.time.Duration
import java.util.*
import java.util.Collections.sort
import java.util.Collections.unmodifiableList


/**
 * @author  Max Dobler - dobler@synyx.de
 */
class Agenda {

    private val timeProvider = Registry.get<TimeProvider>(TimeProvider::class.java)

    private val reservations = ArrayList<Reservation>()

    fun getTimeUntilNextMeeting(): Duration? =
            reservations
                    .filter { it.isUpcoming }
                    .map { it.timeUntilBegin }
                    .firstOrNull()


    fun getTimeUntilAvailable(): Duration =
            if (!hasCurrentReservation()) Duration.ZERO
            else reservations
                    .filter { it.isCurrentOrUpcoming }
                    .map { it.toTimeSpan() }
                    .fold(zeroTimeSpanFrom(timeProvider.now()), sumTimeOfOngoingReservations())
                    .toDuration()


    fun isNextReservationInNearFuture(): Boolean {

        if (hasCurrentReservation()) {
            return false
        }

        val timeUntilNextMeeting = getTimeUntilNextMeeting()

        return timeUntilNextMeeting != null && timeUntilNextMeeting.standardMinutes <= 15L
    }


    fun getCurrentMeeting() = reservations.firstOrNull { it.isCurrentlyActive }


    fun getUpcomingReservation() = getReservationsToday().firstOrNull { it.isUpcoming }


    fun getSecondUpcomingReservation() =
      getReservationsToday()
                    .filter { it.isUpcoming }
                    .drop(1)
                    .firstOrNull()

    fun addReservation(reservation: Reservation) {

        this.reservations.add(reservation)
        sort<Reservation>(reservations)
    }


    fun getReservations(): List<Reservation> = unmodifiableList<Reservation>(reservations)


  fun hasCurrentReservation() = reservations.any { it.isCurrentlyActive }


  private fun sumTimeOfOngoingReservations(): (TimeSpan, TimeSpan) -> TimeSpan {

    return { available, nextTimeSpan ->
      if (nextTimeSpan.begin.isAfter(available.end)) available
      else TimeSpan(available.begin, nextTimeSpan.end)
    }
  }

  public fun getReservationsToday() = reservations.filter { it.isToday }
}
